### PR TITLE
New package: qbs-1.12.0

### PR DESCRIPTION
**Pull request by Cogitri**
qbs does cross compile like this, but I suppose it'd be a good idea to test if it actually does work when cross compiled.

### CODE REVIEW

**Comment by Cogitri** [comment]
This also adds a buildstyle for qbs and updates MellowPlayer, which is built using qbs

File: srcpkgs/mellowplayer/template [hunk 1]
```diff
@@ -1,22 +1,24 @@
 # Template file for 'mellowplayer'
 pkgname=mellowplayer
-version=3.3.5
+version=3.4.0
 revision=1
 wrksrc=MellowPlayer-${version}
-build_style=cmake
-hostmakedepends="pkg-config qt5-tools qt5-qmake"
+build_style=qbs
+configure_args=" config:release"
```
Comment (maxice8) [inline srcpkgs/mellowplayer/template]: is config:release something that needs to be put in all invocations ?

File: common/build-style/qbs.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#
+# This helper is for templates using the Qt Build System (qbs)
+#
+do_configure() {
+	local qmake qt
+	if [ -x "/usr/lib/qt5/bin/qmake" ]; then
```
Comment (maxice8) [inline common/build-style/qbs.sh]: i'm not a big fan of this approach since we can't guarantee that these files won't be brought up as dependencies.

I think it is better we separate into qbs-qt4 and qbs-qt5

Comment (maxice8) [reply common/build-style/qbs.sh]: It does but we can't guarantee the right one will be picked.

With the current script if a dependency pulls in qt-qmake then qt4 will be used unconditionally.

I think it is better we have separate build_style and the developer knows which one to pick an has full confidence the right one will be used

File: srcpkgs/mellowplayer/template (hunk 1 above)
Comment (maxice8) [reply srcpkgs/mellowplayer/template]: If it needs to be put in all invocations like meson is for --buildtype then it is something i'd like in the build_style

Comment (Cogitri) [reply srcpkgs/mellowplayer/template]: > is config:release something that needs to be put in all invocations ?

Yup, otherwise qbs assumes that we've switched from `config:release` to `config:default` and throws the following error (as long as the project in fact doesn't use `config:default` for building):
```
ERROR: Build graph not found for configuration 'default'. Expected location was '/builddir/MellowPlayer-3.4.0/default/default.bg'.
```

> If it needs to be put in all invocations like meson is for --buildtype then it is something i'd like in the build_style

Oh, yeah, this could be done. I'm not quite sure if all qbs packages use this though or if some also use `config:default`

**Comment by Vaelatern** [comment]
I'm a fan of keeping this open but unmerged until we have more than one package that builds this way.

File: common/build-style/qbs.sh (hunk 2 above)
Comment (Cogitri) [reply common/build-style/qbs.sh]: OK, done.

File: srcpkgs/mellowplayer/template (hunk 1 above)
Comment (maxice8) [reply srcpkgs/mellowplayer/template]: Well as @Vaelatern said i think we should wait until more than 1 uses it.

At the moment this build system is too new we can't make much assumptions about it

**Comment by Cogitri** [comment]
> I'm a fan of keeping this open but unmerged until we have more than one package that builds this way.

If we wanted to we could build qt-creator with qbs, but to be honest I don't enjoy using qbs too much, so I'd rather not do that.
https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/qtcreator.qbs

Maybe qt-core and friends will switch from qmake to qbs sometime, but nothing has materialized in their git repos yet and the Qt 5.12 freeze has begun already (so I doubt that we're going to see much qbs adaptation soon)

File: common/build-style/qbs-qt5.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	# makefiles
+	#qbs generate -g makefile ${configure_args}
+}
+
+do_build() {
+	qbs build --no-install ${makejobs} config:release ${configure_args}
+}
+
+do_install() {
+	qbs install --no-build --install-root "$DESTDIR/usr" ${makejobs} \
```
Comment (maxice8) [inline common/build-style/qbs-qt5.sh]: reading the docs we need to pass the -j switch to indicate number of jobs

Comment (Cogitri) [reply common/build-style/qbs-qt5.sh]: Eh? $makejobs already contains a `-j`

**Comment by maxice8** [comment]
On Fri, Sep 07, 2018 at 03:05:21PM -0700, Rasmus Thomsen wrote:
> Cogitri commented on this pull request.
> 
> 
> 
> > +do_configure() {
> +	qbs-setup-qt /usr/lib/qt5/bin/qmake qt5
> +	qbs-config defaultProfile qt5
> +
> +	# In theory qbs supports generating makefiles, however most projects
> +	# who utilize qbs utilize behavior qbs apparently can't use when using
> +	# makefiles
> +	#qbs generate -g makefile ${configure_args}
> +}
> +
> +do_build() {
> +	qbs build --no-install ${makejobs} config:release ${configure_args}
> +}
> +
> +do_install() {
> +	qbs install --no-build --install-root "$DESTDIR/usr" ${makejobs} \
> 
> Eh? $makejobs already contains a `-j`
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/pull/2503#discussion_r216098439

oh, must have confused with XBPS_MAKEJOBS

File: common/environment/build-style/qbs-qt4.sh [hunk 4]
```diff
@@ -0,0 +1 @@
+hostmakedepends+=' qbs'
```
Comment (maxice8) [inline common/environment/build-style/qbs-qt4.sh]: double quotes please and i think adding qmake-qt is safe, same for qbs-qt5

File: srcpkgs/mellowplayer/template [hunk 5]
```diff
@@ -1,22 +1,23 @@
 # Template file for 'mellowplayer'
 pkgname=mellowplayer
-version=3.3.5
+version=3.4.0
 revision=1
 wrksrc=MellowPlayer-${version}
-build_style=cmake
-hostmakedepends="pkg-config qt5-tools qt5-qmake"
```
Comment (Johnnynator) [inline srcpkgs/mellowplayer/template]: you need `qbs` and `qt5-qmake` either here in hostmakedepends or add them in `common/environment/build-style/qbs-qt5.sh` Also I would be surprised if the qbs-qt5 build-style actually cross compiles.

Comment (Cogitri) [reply srcpkgs/mellowplayer/template]: Yup, cross compiling doesn't work as of now - but QBS docs on cross compiling seem a little "suboptimal" to me (or I just don't find the right ones), so I'm not quite sure how I would get it to cross compile.

Comment (Johnnynator) [reply srcpkgs/mellowplayer/template]: Adding a cross toolchain should be fairly straight forward e.g. `qbs setup-toolchains aarch64-linux-gnu-gcc aarch64 `. The biggest problem is `qmake` or better `qbs-setup-qt /usr/lib/qt5/bin/qmake qt5` since qmake ( `qmake-qt5 -query` ) will always return the host configuration (which is AFAIK also a problem in qmake build-style). In my understanding of qmake (or lack) we would need to compile qmake for every target architecture we want to cross-build for to feed qbs with the correct path, but I highly suspect that this would break other stuff.

TL;DR qmake sucks and qbs depends on it.

Comment (Cogitri) [reply srcpkgs/mellowplayer/template]: > Adding a cross toolchain should be fairly straight forward e.g. qbs setup-toolchains aarch64-linux-gnu-gcc aarch64

Yup, I've done that.

> The biggest problem is qmake or better qbs-setup-qt /usr/lib/qt5/bin/qmake qt5 since qmake ( qmake-qt5 -query ) will always return the host configuration (which is AFAIK also a problem in qmake build-style).

If I understand correctly qbs offers a `sysroot` option (and so does qmake), so if we were to set `sysroot` to /usr/$XBPS_CROSS_TRIPLET it may work. 

> In my understanding of qmake (or lack) we would need to compile qmake for every target architecture we want to cross-build for to feed qbs with the correct path

Uhh..how would we run that cross compiled qmake on the host? qbs runs `qmake -query` to get Qt's paths.

> TL;DR qmake sucks and qbs depends on it.

Agreed.

Comment (Cogitri) [reply srcpkgs/mellowplayer/template]: > One can always compile it natively.

Sure, I like my package manager to do that though and to upstream my work, so I guess I'm stuck with this

Comment (travankor) [reply srcpkgs/mellowplayer/template]: Last I heard, QT is not planning switching build systems until QT 6, which will first appear in late 2020.

> > One can always compile it natively.
> 
> Sure, but I like my package manager to do that though and to upstream my work, so I guess I'm stuck with this

Not ideal, but there seems to be an appimage available for the latest mellow player.

BTW why did mellow player guy switch cold turkey to a build system that no one uses? At the very least, he could of kept the qmake option.

Comment (Cogitri) [reply srcpkgs/mellowplayer/template]: > Not ideal, but there seems to be an appimage available for the latest mellow player.

Yeah, not a huge fan of outdated libs :)

>  BTW why did mellow player guy switch cold turkey to a build system that no one uses? At the very least, he could of kept the qmake option.

Uhh...MellowPlayer never offered a QMake option, previously it used CMake for building. But apparently it's easier for the MellowPlayer maintainer to use qbs because it's a little nicer to use for Qt projects than CMake (but hell for everyone else)

Comment (travankor) [reply srcpkgs/mellowplayer/template]: err right.. in any case, cmake is a lot nicer than qmake or qbs from the end-user perspective.

But who knows... maybe QT will adopt meson and deprecate qbs

Comment (Cogitri) [reply srcpkgs/mellowplayer/template]: Yep, CMake (or meson!) would be way nicer, but oh well. I very much doubt that Qt won't use qbs, why would they write an entire build system only to not use it? This might not be the right place to discuss this though

**Comment by nonchip** [comment]
> I'm a fan of keeping this open but unmerged until we have more than one package that builds this way.

i'm a fan of merging this at least qbs because it doesn't matter how many things on this repo build with it, other things already do. i mean how are we supposed to build either 3rd party stuff OR new packages using it without access to this build system?

i mean yeah i totally get the stuff about the outdated mellow stuff but if you apply that to qbs just because you dont want mellow i'm afraid i have to ask: when did the void philosophy change from "have a package you'd like to get included? give us a PR so we don't get that AUR crap!" to "uh i dont know enough stuff that I'd use this for, no thanks"?

**Comment by Cogitri** [comment]
> then again those packages really shouldnt be in the same PR, that's just unmaintainable bad style.

MellowPlayer is in this PR because it's the only package that uses QBS in Void's repos as of now and is meant as a test case for qbs.

qbs needs fixing for cross compiling anyway - if you want to you can do that, the less time I have to spend with qbs the better for my sanity.

**Comment by nonchip** [comment]
> MellowPlayer is in this PR because it's the only package that uses qbs in Void's repos as of now and is meant as a test case for qbs.

i see your logic, but that's not how it's supposed to work here. the PR is about "New Package: qbs" so it should be only a new package for qbs. which also would increase the likelihood of it being merged. also the contributing guidelines make it very clear that this is a bad practice so it's likely this PR won't ever be merged in the current form. i even got PRs declined for the simple fact they contained 2 commits instead of one since i fixed a typo, so yeah i don't think this one will be considered "maintainable enough". might feel a bit annoying to have all those rules but without them it'd just be a bugged anarchy i guess :P

> qbs needs fixing for cross compiling anyway - if you want to you can do that, the less time I have to spend with qbs the better for my sanity.

sadly i have no real clue about cross compiling (except for `x64 -> avr or z80` or clicking the "make HTML5" button in unreal engine, but i don't think either of that helps in this case), i wouldn't even know where to start.

**Comment by Hoshpak** [comment]
Qbs is now [officially deprecated](https://blog.qt.io/blog/2018/10/29/deprecation-of-qbs/).

**Comment by maxice8** [comment]
Thank God

On Oct 29, 2018 15:46, "Helmut Pozimski" <notifications@github.com> wrote:

Qbs is now officially deprecated
<https://blog.qt.io/blog/2018/10/29/deprecation-of-qbs/>.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/void-linux/void-packages/pull/2503#issuecomment-434031926>,
or mute the thread

**Comment by Cogitri** [comment]
Yup, a already notified Upstream, they're yoing to switch back to CMakw
